crypto/ec/ec_mult.c

```diff
@@ -108,62 +108,83 @@ void EC_ec_pre_comp_free(EC_PRE_COMP *pre)
+ * `scalar` cannot be NULL and should be in the range [0,n) otherwise all
+ * constant time bets are off (where n is the cardinality of the EC group).
+ *
+ * NB: This says nothing about the constant-timeness of the ladder step
+ * implementation (i.e., the default implementation is based on EC_POINT_add and
+ * EC_POINT_dbl, which of course are not constant time themselves) or the
+ * underlying multiprecision arithmetic.
  *
- * NB: This says nothing about EC_POINT_add and EC_POINT_dbl,
- * which of course are not constant time themselves.
+ * The product is stored in `r`.
  *
- * The product is stored in r.
+ * This is an internal function: callers are in charge of ensuring that the
+ * input parameters `group`, `r`, `scalar` and `ctx` are not NULL.
  *
  * Returns 1 on success, 0 otherwise.
  */
-static int ec_mul_consttime(const EC_GROUP *group, EC_POINT *r,
-                            const BIGNUM *scalar, const EC_POINT *point,
-                            BN_CTX *ctx)
+int ec_scalar_mul_ladder(const EC_GROUP *group, EC_POINT *r,
+                         const BIGNUM *scalar, const EC_POINT *point,
+                         BN_CTX *ctx)
 {
     int i, cardinality_bits, group_top, kbit, pbit, Z_is_one;
+    EC_POINT *p = NULL;
     EC_POINT *s = NULL;
     BIGNUM *k = NULL;
     BIGNUM *lambda = NULL;
     BIGNUM *cardinality = NULL;
-    BN_CTX *new_ctx = NULL;
     int ret = 0;
 
-    if (ctx == NULL && (ctx = new_ctx = BN_CTX_secure_new()) == NULL)
-        return 0;
+    /* early exit if the input point is the point at infinity */
+    if (point != NULL && EC_POINT_is_at_infinity(group, point))
+        return EC_POINT_set_to_infinity(group, r);
 
     BN_CTX_start(ctx);
 
-    s = EC_POINT_new(group);
-    if (s == NULL)
+    if (((p = EC_POINT_new(group)) == NULL)
+        || ((s = EC_POINT_new(group)) == NULL)) {
+        ECerr(EC_F_EC_SCALAR_MUL_LADDER, ERR_R_MALLOC_FAILURE);
         goto err;
+    }
 
     if (point == NULL) {
-        if (!EC_POINT_copy(s, group->generator))
+        if (!EC_POINT_copy(p, group->generator)) {
+            ECerr(EC_F_EC_SCALAR_MUL_LADDER, ERR_R_EC_LIB);
             goto err;
+        }
     } else {
-        if (!EC_POINT_copy(s, point))
+        if (!EC_POINT_copy(p, point)) {
+            ECerr(EC_F_EC_SCALAR_MUL_LADDER, ERR_R_EC_LIB);
             goto err;
+        }
     }
 
+    EC_POINT_BN_set_flags(p, BN_FLG_CONSTTIME);
+    EC_POINT_BN_set_flags(r, BN_FLG_CONSTTIME);
     EC_POINT_BN_set_flags(s, BN_FLG_CONSTTIME);
 
     cardinality = BN_CTX_get(ctx);
     lambda = BN_CTX_get(ctx);
     k = BN_CTX_get(ctx);
-    if (k == NULL || !BN_mul(cardinality, group->order, group->cofactor, ctx))
+    if (k == NULL) {
+        ECerr(EC_F_EC_SCALAR_MUL_LADDER, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    if (!BN_mul(cardinality, group->order, group->cofactor, ctx)) {
+        ECerr(EC_F_EC_SCALAR_MUL_LADDER, ERR_R_BN_LIB);
```

Comment:
One might argue that this case might be worthy of introducing a new "reason" code, e.g.,  `EC_R_CANNOT_COMPUTE_CARDINALITY` or something similar.

What do you think about it?

Reply:
@romen Given that after #6648 we learned defaults for order and cofactor are not NULL but 0, I would argue a very early abort `BN_is_zero(group->order) || BN_is_zero(group->cardinality)` would be more appropriate.

In which case this particular `BN_mul` failure is pretty generic.

Reply:
I was doing this, and deciding on the proper reason code and discovered that the order could officially be INVALID, UNDEFINED, UNKNOWN or WRONG:

```bash
~/repos/openssl >>> ack -B1 'EC_R_.*_ORDER' **/*.c
crypto/ec/ec_asn1.c
781-    if (BN_is_negative(a) || BN_is_zero(a)) {
782:        ECerr(EC_F_EC_GROUP_NEW_FROM_ECPARAMETERS, EC_R_INVALID_GROUP_ORDER);
--
785-    if (BN_num_bits(a) > (int)field_bits + 1) { /* Hasse bound */
786:        ECerr(EC_F_EC_GROUP_NEW_FROM_ECPARAMETERS, EC_R_INVALID_GROUP_ORDER);

crypto/ec/ec_check.c
54-    if (BN_is_zero(order)) {
55:        ECerr(EC_F_EC_GROUP_CHECK, EC_R_UNDEFINED_ORDER);
--
61-    if (!EC_POINT_is_at_infinity(group, point)) {
62:        ECerr(EC_F_EC_GROUP_CHECK, EC_R_INVALID_GROUP_ORDER);

crypto/ec/ec_err.c
313-    {ERR_PACK(ERR_LIB_EC, 0, EC_R_INVALID_FORM), "invalid form"},
314:    {ERR_PACK(ERR_LIB_EC, 0, EC_R_INVALID_GROUP_ORDER), "invalid group order"},
--
356-    {ERR_PACK(ERR_LIB_EC, 0, EC_R_UNDEFINED_GENERATOR), "undefined generator"},
357:    {ERR_PACK(ERR_LIB_EC, 0, EC_R_UNDEFINED_ORDER), "undefined order"},
358-    {ERR_PACK(ERR_LIB_EC, 0, EC_R_UNKNOWN_GROUP), "unknown group"},
359:    {ERR_PACK(ERR_LIB_EC, 0, EC_R_UNKNOWN_ORDER), "unknown order"},
--
362-    "wrong curve parameters"},
363:    {ERR_PACK(ERR_LIB_EC, 0, EC_R_WRONG_ORDER), "wrong order"},

crypto/ec/ec_key.c
299-    if (BN_is_zero(order)) {
300:        ECerr(EC_F_EC_KEY_SIMPLE_CHECK_KEY, EC_R_INVALID_GROUP_ORDER);
--
307-    if (!EC_POINT_is_at_infinity(eckey->group, point)) {
308:        ECerr(EC_F_EC_KEY_SIMPLE_CHECK_KEY, EC_R_WRONG_ORDER);
--
316-        if (BN_cmp(eckey->priv_key, order) >= 0) {
317:            ECerr(EC_F_EC_KEY_SIMPLE_CHECK_KEY, EC_R_WRONG_ORDER);

crypto/ec/ec_mult.c
835-    if (BN_is_zero(order)) {
836:        ECerr(EC_F_EC_WNAF_PRECOMPUTE_MULT, EC_R_UNKNOWN_ORDER);

crypto/ec/ecp_nistz256.c
827-    if (BN_is_zero(order)) {
828:        ECerr(EC_F_ECP_NISTZ256_MULT_PRECOMPUTE, EC_R_UNKNOWN_ORDER);
```

For consistency with `ec_mult.c` and `ecp_nistz256.c` I am going with `EC_R_UNKNOWN_ORDER` (and adding a similar one for the cofactor), but to me it looks like `EC_R_UNDEFINED_ORDER` from `EC_GROUP_check` has exactly the same semantic, and it's a bit confusing!